### ISSUE TITLE
Rule Proposal: `arrow-block-style`

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#4109]
This rule is a styling rule for "non block" v.s. "single return block".

There is a trap in arrow function that returns an object.

```js
let foo = () => {};
```

This `{}` is a block, not an empty object.
This should be replaced with:

```js
let foo = () => ({});
// or
let foo = () => { return {}; };
// or
let foo = () => undefined;
// or
let foo = () => { /* do nothing */ };
```

This rule has an option.

- `"as-needed"` (default) - If it's possible, this suggests a use of expression without a block.
- `"always"` - This enforces using a block. 

----

```js
{
    "arrow-block-style": [2, "as-needed"]
}
```

**Valid**

```js
// OK, this is not a block.
let foo = () => 0;

// OK, this cannot replace with non-block (without a comma-sequence).
let foo = (retv, name) => {
    retv[name] = true;
    return retv;
};

// OK, this is considered as having some side effects, so the block is proper.
let foo = () => { bar(); };

// OK, it's explicit functions that does nothing.
// This behavior is the same as http://eslint.org/docs/rules/no-empty
let foo = () => { /* do nothing */ };
let foo = () => {
    // do nothing.
};
```

**Invalid**

```js
// This can be replaced with `() => 0` simply.
let foo = () => {
    return 0;
};

// This `{}` is a block.
// it should be fixed to `() => ({})` or `() => undefined` or `() => { /* do nothing */ }`.
let foo = () => {};
```

----

```js
{
    "arrow-block-style": [2, "always"]
}
```

**Valid**

```js
let foo = () => {
    return 0;
};
let foo = (retv, name) => {
    retv[name] = true;
    return retv;
};
```

**Invalid**

```js
let foo = () => 0;
```

----

**Comment by alberto** [comment]
@mysticatea are you working on this?

**Comment by BenoitZugmeyer** [comment]
Awesome. I would love to see a third option, `as-needed-line`, in the same spirit of `curly multi-line`, enforcing one-liners to behave like `as-needed` and multi-line functions like `always`.

```js
{
    "arrow-block-style": [2, "as-needed"]
}
```

##### Valid

```js
// OK, one liner, valid
let foo = () => 0;

// OK, cannot be replaced by non-block
let foo = (retv, name) => {
    retv[name] = true;
    return retv;
};

// OK, this is considered as having some side effects, so the block is proper.
let foo = () => { bar(); };

// OK, multiline
let foo = () => {
    return 0;
};
```

##### Invalid

```js
// Invalid, multiline
let foo = () =>
    0;

// Invalid, multiline
let foo = () => foo(
    true,
    false);
```

Should I open a new issue for this?

**Comment by mysticatea** [comment]
@alberto I'm not working on this :)

@BenoitZugmeyer I'd like to separate issue. But the rule does not exist yet. Umm...

**Comment by BenoitZugmeyer** [comment]
OK no problem, I'll open an issue when this rule is implemented.

**Comment by alberto** [comment]
Ok, I have the examples mentioned here covered, but I'm not sure all potential cases are.
For example, this is also probably an error:
```
var foo = () => {bar: 1};
```

Should  this be an error too?
```
var foo = () => { a };
```

**Comment by mysticatea** [comment]
Wow, thank you all.

@alberto Good point. I am not sure this rule should catch those pattern since those will be caught by other rules (e.g. no-unused-expression). Well, is there a rule which allows only loop in LabeledStatement? (`{foo: bar()}` is caught by other?) I want to hear advices. @eslint/eslint-team

**Comment by hzoo** [comment]
Right now there's http://eslint.org/docs/rules/no-labels, no-label-var, no-empty-label. I don't think there's a rule that only allows labels in a loop

**Comment by mysticatea** [comment]
@hzoo Thank you! 

[no-empty-label](https://github.com/eslint/eslint/blob/master/lib/rules/no-empty-label.js) seems to catch `() => {foo: bar()}`, so in my opinion, special handling is not needed for those two pattern. I guess those are handled in the same path as `let foo = () => { bar(); };` (the block does not have one ReturnStatement).
But my opinion is not strong.

**Comment by alberto** [comment]
In which category should this rule fall under?

**Comment by alberto** [comment]
Nevermind, I just saw there is a specific category for ES6 features.

**Comment by nzakas** [comment]
Yeah, I don't think we need this to catch any further patterns, as we're covered by other rules.

**Comment by alberto** [comment]
Can anyone have a look at this?
